File: tests/test-minecraft/src/main/java/pize/tests/minecraft/game/resources/TextureResource.java
package pize.tests.minecraft.game.resources;

import pize.graphics.texture.Texture;
import pize.graphics.texture.TextureRegion;
import pize.tests.minecraft.utils.log.Logger;

import java.awt.*;

public class TextureResource extends Resource<TextureRegion>{

    private TextureRegion texture;
    private Rectangle region;
    private boolean loaded;

    public TextureResource(String location,Rectangle region){
        super(location);

        this.region = region;
    }

    public TextureResource(String location){
        this(location,null);
    }

    @Override
    public void loadResource(){
        try{
            if(region == null)
                texture = new TextureRegion(new Texture(getLocation()));
            else
                texture = new TextureRegion(new Texture(getLocation()), (int) region.getX(), (int) region.getY(), (int) region.getWidth(), (int) region.getHeight());

            loaded = true;
        }catch(RuntimeException e){
            loaded = false;
            Logger.instance().warn(e.getLocalizedMessage());
        }
    }

    @Override
    public void reloadResource(){
        TextureRegion oudTexture = texture;
        loadResource();
        oudTexture.texture.dispose();
    }


    public void setRegion(Rectangle region){
        this.region = region;
    }


    @Override
    public boolean isLoaded(){
        return loaded;
    }

    @Override
    public TextureRegion getResource(){
        return texture;
    }

    @Override
    public void dispose(){
        texture.texture.dispose();
    }

}

File: tests/test-minecraft/src/main/kotlin/pize/tests/minecraft/game/resources/TextureResource.kt
package pize.tests.minecraft.game.resources

import pize.graphics.texture.Texture
import pize.graphics.texture.TextureRegion
import pize.tests.minecraft.utils.log.Logger
import java.awt.Rectangle

class TextureResource @JvmOverloads constructor(location: String, private var region: Rectangle? = null) :
    Resource<TextureRegion?>(location) {
    override var resource: TextureRegion? = null
        private set
    override var isLoaded = false
        private set

    override fun loadResource() {
        try {
            if (region == null) resource = TextureRegion(Texture(location)) else resource = TextureRegion(
                Texture(
                    location
                ),
                region!!.getX().toInt(),
                region!!.getY().toInt(),
                region!!.getWidth().toInt(),
                region!!.getHeight().toInt()
            )
            isLoaded = true
        } catch (e: RuntimeException) {
            isLoaded = false
            Logger.Companion.instance().warn(e.localizedMessage)
        }
    }

    override fun reloadResource() {
        val oudTexture = resource
        loadResource()
        oudTexture!!.texture!!.dispose()
    }

    fun setRegion(region: Rectangle?) {
        this.region = region
    }

    override fun dispose() {
        resource!!.texture!!.dispose()
    }
}
